pkg/operator/operator.go

```diff
@@ -67,68 +63,62 @@ type Operator struct {
 }
 
 func NewOperator(ctx context.Context, operator *operator.Operator) (context.Context, *Operator) {
-	azConfig, err := GetAZConfig()
-	lo.Must0(err, "creating Azure config") // NOTE: we prefer this over the cleaner azConfig := lo.Must(GetAzConfig()), as when initializing the client there are helpful error messages in initializing clients and the azure config
+	opts := options.FromContext(ctx)
 
-	azClient, err := instance.CreateAZClient(ctx, azConfig)
-	lo.Must0(err, "creating Azure client")
+	var authManager *auth.AuthManager
+	if opts.ArmAuthMethod == auth.AuthMethodWorkloadIdentity {
+		authManager = auth.NewAuthManagerWorkloadIdentity(opts.Location)
+	} else if opts.ArmAuthMethod == auth.AuthMethodSysMSI {
+		authManager = auth.NewAuthManagerSystemAssignedMSI(opts.Location)
+	}
 
-	vnetGUID, err := getVnetGUID(azConfig, options.FromContext(ctx).SubnetID)
-	lo.Must0(err, "getting VNET GUID")
+	azClient, err := instance.CreateAZClient(ctx, opts.SubscriptionID, authManager)
+	lo.Must0(err, "creating Azure client")
 
 	unavailableOfferingsCache := azurecache.NewUnavailableOfferings()
 	pricingProvider := pricing.NewProvider(
 		ctx,
+		opts,
 		pricing.NewAPI(),
-		azConfig.Location,
 		operator.Elected(),
 	)
 	imageProvider := imagefamily.NewProvider(
+		opts,
 		operator.KubernetesInterface,
 		cache.New(azurecache.KubernetesVersionTTL,
 			azurecache.DefaultCleanupInterval),
 		azClient.ImageVersionsClient,
-		azConfig.Location,
 	)
 	imageResolver := imagefamily.New(
 		operator.GetClient(),
 		imageProvider,
 	)
 	launchTemplateProvider := launchtemplate.NewProvider(
 		ctx,
+		opts,
```

Comment:
This should be the highlights worth discussing.

Prior to this, in addition to passing the options (and `azConfig`) directly that you can see in the changes preview, some of the providers did call `options.FromContext(ctx)` again. I find this inconsistent and would prefer to stick to one side—either pass all required parameters here, or just pass in `ctx`/`opts`.

The current code chooses passing in `opts`. This will make the providers depends on the knowledge of options (no change), while sparing operator from having to know which provider needs what. This is particularly helpful in `launchTemplateProvider`, which is have a lot of demands for fields in options.

With that said, each provider have different demands in options fields, but ideally I want to keep the dependencies/pattern consistent for all providers.

Reply:
Alternatively, we could stay explicit on the interface and pass all variables that we currently getting from `options.FromContext(ctx)` through here instead. The interface will be be big for high-demanding providers, but we will be able to cut their dependence on `options`.

Reply:
Another alternative is to add some layering in `options` to categorize them by uses, appearing at "sub-structs".
But this will force `options` to contain knowledge about the exact uses of their fields, which is going to add additional complexity, given some fields are shared with multiple uses.

An extension to that is to have additional layer between `options` (which will mostly take envs in raw) and this layer to map those to its provider uses. But that would be no difference that just creating structs here, which will only be better than explicit passing by extensibility and small reusability, at the cost of additional complexity.

Reply:
But all these options are not really making use of `options.FromContext(ctx)` pattern established by karpenter-core and AWS provider. If we want to keep that, another options is to just pass in `ctx` to all places and let `options.FromContext(ctx)` retrieve the options whenever it is needed. This will assume that all instances here are singletons or at least have no differences between each, which is true(?), for now.